pkg/proxy/healthcheck/proxier_health.go

```diff
@@ -78,77 +79,101 @@ func NewProxierHealthServer(addr string, healthTimeout time.Duration) ProxierHea
-func (hs *proxierHealthServer) Updated() {
-	hs.oldestPendingQueued.Store(zeroTime)
-	hs.lastUpdated.Store(hs.clock.Now())
+// Updated indicates that proxier for the given IP family has successfully updated
+// its backend, so it should be considered healthy now.
+func (hs *proxierHealthServer) Updated(ipFamily v1.IPFamily) {
+	hs.lock.Lock()
+	defer hs.lock.Unlock()
+	hs.oldestPendingQueuedMap[ipFamily] = zeroTime
+	hs.lastUpdatedMap[ipFamily] = hs.clock.Now()
 }
 
-// QueuedUpdate indicates that the proxy has received changes from the apiserver but
-// has not yet pushed them to its backend. If the proxy does not call Updated within the
-// healthTimeout time then it will be considered unhealthy.
-func (hs *proxierHealthServer) QueuedUpdate() {
-	// Set oldestPendingQueued only if it's currently zero
-	hs.oldestPendingQueued.CompareAndSwap(zeroTime, hs.clock.Now())
+// QueuedUpdate indicates that the proxier for the given IP family has received changes
+// from the APIServer but has not yet pushed them to its backend. If the proxy does not
+// call Updated within the healthTimeout time then it will be considered unhealthy.
+func (hs *proxierHealthServer) QueuedUpdate(ipFamily v1.IPFamily) {
+	hs.lock.Lock()
+	defer hs.lock.Unlock()
+	// Set oldestPendingQueuedMap only if it's currently zero
+	if val, ok := hs.oldestPendingQueuedMap[ipFamily]; ok && val == zeroTime {
+		hs.oldestPendingQueuedMap[ipFamily] = hs.clock.Now()
+	}
 }
 
 // IsHealthy returns only the proxier's health state, following the same
 // definition the HTTP server defines, but ignoring the state of the Node.
 func (hs *proxierHealthServer) IsHealthy() bool {
-	isHealthy, _, _ := hs.isHealthy()
+	isHealthy, _ := hs.isHealthy()
 	return isHealthy
 }
 
-func (hs *proxierHealthServer) isHealthy() (bool, time.Time, time.Time) {
-	var oldestPendingQueued, lastUpdated time.Time
-	if val := hs.oldestPendingQueued.Load(); val != nil {
-		oldestPendingQueued = val.(time.Time)
-	}
-	if val := hs.lastUpdated.Load(); val != nil {
-		lastUpdated = val.(time.Time)
-	}
+func (hs *proxierHealthServer) isHealthy() (bool, time.Time) {
+	hs.lock.RLock()
+	defer hs.lock.RUnlock()
+
+	var lastUpdated, proxierLastUpdated time.Time
 	currentTime := hs.clock.Now()
 
-	healthy := false
-	switch {
-	case oldestPendingQueued.IsZero():
-		// The proxy is healthy while it's starting up
-		// or the proxy is fully synced.
-		healthy = true
-	case currentTime.Sub(oldestPendingQueued) < hs.healthTimeout:
-		// There's an unprocessed update queued, but it's not late yet
-		healthy = true
+	for ipFamily, oldestPendingQueued := range hs.oldestPendingQueuedMap {
+		if _, ok := hs.lastUpdatedMap[ipFamily]; ok {
+			proxierLastUpdated = hs.lastUpdatedMap[ipFamily]
+			if proxierLastUpdated.After(lastUpdated) {
+				lastUpdated = proxierLastUpdated
+			}
+		}
+
+		switch {
+		case oldestPendingQueued.IsZero():
+			// The proxier is healthy while it's starting up
+			// or the proxier is fully synced.
+			continue
```

Comment:
So again, this case would never get hit. The initial state is that there's no entry for that IP family, not a zero entry. But we don't need any special case for this here now; when the proxy is starting up (ie, `oldestPendingQueuedMap` is empty), the `for` loop just gets skipped and we return `true` below.